File: kernel/src/mem/address_space.cpp
#include "address_space.h"
#include "pmm.h"
#include "vmm.h"
#include "x86/page_table.h"
#include "x86/asm.h"

bool AddressSpace::is_user_address(uintptr_t addr) {
	return addr < 0x800000000000;
}

bool AddressSpace::is_user_range(uintptr_t addr, size_t len) {
	if (addr + len < addr)
		return false;
	return is_user_address(addr) && is_user_address(addr + len - 1);
}

bool AddressSpace::is_user_range(const Range& range) {
	return is_user_range(range.base(), range.size());
}

AddressSpace::AddressSpace(bool create)
	: m_page_table(create)
{
	if (!create)
		m_page_table.set_current_cr3();
}

// AddressSpace::AddressSpace(uintptr_t ptl4_paddr)
// 	: m_page_table(ptl4_paddr)
// {
// }

uintptr_t AddressSpace::find_free_memory_region_for_range(const Range& range) {
	ASSERT(range.base() == 0, "range already has base address");
	uintptr_t base = USER_MAPPINGS_START_ADDR;
	while (is_user_address(base)) {
		uintptr_t length;
		for (length = 0; length < range.length(); length += PAGE_SIZE) {
			PageTableEntry* pte = m_page_table.ensure_pte(base + length);
			if (!pte)
				return 0;
			if (pte->is_present())
				break;
		}

		// If the loop got to the end, we found our desired region
		if (length == range.length())
			return base;

		base += length + PAGE_SIZE;
	}
	return 0;
}

bool AddressSpace::map_range(Range& range, uint8_t perms, uint8_t flags) {
	// Assign a base address to the range in case it doesn't have one
	if (!range.base()) {
		uintptr_t base = find_free_memory_region_for_range(range);
		if (!base) {
			return false;
		}
		range.set_base(base);
	}

	// Check the range is in user range
	if (!is_user_range(range)) {
		return false;
	}

	// Attempt to allocate physical memory for the range
	vector<uintptr_t> frames;
	size_t num_frames = PAGE_CEIL(range.size()) / PAGE_SIZE;
	if (!PMM::alloc_frames(num_frames, frames)) {
		return false;
	}

	// Compute page flags according to memory permissions and mapping flags
	uint64_t page_flags = PageTableEntry::User;
	if (perms == MemPerms::None) {
		// ProtNone (which is the same as Global) without Present
		page_flags |= PageTableEntry::ProtNone;
	} else {
		page_flags |= PageTableEntry::Present;
		if (perms & MemPerms::Write)
			page_flags |= PageTableEntry::ReadWrite;
		if (!(perms & MemPerms::Exec))
			page_flags |= PageTableEntry::NoExecute;
	}
	if (flags & MapFlags::Shared)
		page_flags |= PageTableEntry::Shared;
	bool discard = flags & MapFlags::DiscardAlreadyMapped;

	// Map every page
	for (size_t i = 0; i < frames.size(); i++) {
		uintptr_t page_base = range.base() + i*PAGE_SIZE;
		if (!m_page_table.map(page_base, frames[i], page_flags, discard)) {
			return false;
		}
	}
	return true;
}

bool AddressSpace::unmap_range(const Range& range, bool ignore_not_mapped) {
	// Check the range is in user range
	if (!is_user_range(range))
		return false;

	// Unmap every page
	for (uintptr_t page_base = range.base();
	     page_base < range.base() + range.size();
	     page_base += PAGE_SIZE)
	{
		if (!m_page_table.unmap(page_base, ignore_not_mapped))
			return false;
	}
	return true;
}

bool AddressSpace::set_range_perms(const Range& range, uint8_t perms) {
	// Check the range is in user range
	if (!is_user_range(range))
		return false;

	// Set permissions to every page
	PageTableEntry* pte;
	for (uintptr_t page_base = range.base();
	     page_base < range.base() + range.size();
	     page_base += PAGE_SIZE)
	{
		pte = m_page_table.ensure_pte(page_base);
		if (!pte) {
			return false;
		}
		if (!pte->is_present()) {
			return false;
		}
		pte->set_prot_none(perms == MemPerms::None);
		pte->set_writable(perms & MemPerms::Write);
		pte->set_execute_disabled(!(perms & MemPerms::Exec));
		flush_tlb_entry(page_base);
	}
	return true;
}

Optional<AddressSpace> AddressSpace::clone() const {
	auto page_table_copy = m_page_table.clone();
	if (!page_table_copy.has_value())
		return false;

	AddressSpace copy(false);
	copy.m_page_table = page_table_copy.value();
	return copy;
}

void AddressSpace::load() {
	m_page_table.load();
}

bool AddressSpace::operator==(const AddressSpace& other) const {
	return m_page_table == other.m_page_table;
}

bool AddressSpace::operator!=(const AddressSpace& other) const {
	return !(*this == other);
}

File: kernel/src/mem/address_space.zig
usingnamespace @import("../common.zig");
const mem = @import("mem.zig");
const x86 = @import("../x86/x86.zig");
const PageTable = x86.paging.PageTable;
const Allocator = std.mem.Allocator;

pub const Perms = packed struct {
    read: bool = false,
    write: bool = false,
    exec: bool = false,

    pub fn isNone(perms: Perms) bool {
        return !perms.read and !perms.write and !perms.exec;
    }
};

pub const AddressSpace = struct {
    allocator: *Allocator,
    page_table: PageTable,

    pub fn fromCurrent(allocator: *Allocator) AddressSpace {
        return AddressSpace{
            .allocator = allocator,
            .page_table = PageTable.fromCurrent(),
        };
    }

    pub fn load(self: *AddressSpace) void {
        self.page_table.load();
    }

    pub const MapFlags = packed struct {
        discardAlreadyMapped: bool = false,
        shared: bool = false,
    };

    fn checkRange(addr: usize, length: usize) !void {
        assert(mem.isPageAligned(addr));
        assert(mem.isPageAligned(length));
        assert(length != 0);
        assert(!std.meta.isError(std.math.add(usize, addr, length)));
        if (!mem.safe.isRangeInUserRange(addr, length))
            return error.NotUserRange;
    }

    const MappingError = PageTable.MappingError || error{NotUserRange};

    // Can't return error.AlreadyMapped if flags.discardAlreadyMapped is set.
    pub fn mapRange(
        self: *AddressSpace,
        addr: usize,
        length: usize,
        perms: Perms,
        flags: MapFlags,
    ) MappingError!void {
        try checkRange(addr, length);

        // Attempt to allocate physical memory for the range
        const num_frames = @divExact(length, std.mem.page_size);
        var frames = try mem.pmm.allocFrames(self.allocator, num_frames);
        defer self.allocator.free(frames);

        // Get the mapping options acording to memory permissions and mapping flags
        const mapping_options = x86.paging.PageTable.MappingOptions{
            .writable = perms.write,
            .user = true,
            .protNone = perms.isNone(),
            .shared = flags.shared,
            .noExecute = !perms.exec,
            .discardAlreadyMapped = flags.discardAlreadyMapped,
        };

        // Map every page
        var i: usize = 0;
        var page_base: usize = addr;
        errdefer {
            // We won't unmap the mapped pages, as that's what sys_mmap requires,
            // so just free the unused frames
            mem.pmm.freeFrames(frames[i..]);
        }
        while (i < num_frames) : ({
            i += 1;
            page_base += std.mem.page_size;
        }) {
            try self.page_table.mapPage(page_base, frames[i], mapping_options);
        }
    }

    pub fn mapRangeAnywhere(
        self: *AddressSpace,
        length: usize,
        perms: Perms,
        flags: MapFlags,
    ) error{OutOfMemory}!usize {
        assert(mem.isPageAligned(length));
        if (self.findFreeRange(length)) |range_base_addr| {
            self.mapRange(range_base_addr, length, perms, flags) catch |err| switch (err) {
                error.AlreadyMapped, error.NotUserRange => unreachable,
                error.OutOfMemory => return error.OutOfMemory,
            };
            return range_base_addr;
        }
        return error.OutOfMemory;
    }

    fn findFreeRange(self: *AddressSpace, range_length: usize) ?usize {
        var base = mem.layout.user_mappings_start;
        while (mem.safe.isAddressInUserRange(base)) {
            var length: usize = 0;
            while (length < range_length and !self.page_table.isMapped(base + length)) {
                length += std.mem.page_size;
            }

            // If the loop got to the end, we found our desired region
            if (length == range_length)
                return base;

            base += length + std.mem.page_size;
        }
        return null;
    }

    const UnmappingError = PageTable.UnmappingError || error{NotUserRange};

    /// Unmap every mapped page in a range of memory. If there's any not mapped
    /// page, it will be ignored, and at the end it will return error.NotMapped.
    pub fn unmapRange(
        self: *AddressSpace,
        addr: usize,
        length: usize,
    ) UnmappingError!void {
        try checkRange(addr, length);

        // Unmap every page
        var not_mapped: bool = false;
        const addr_end = addr + length;
        var page_base: usize = addr;
        while (page_base < addr_end) : (page_base += std.mem.page_size) {
            self.page_table.unmapPage(page_base) catch |err| switch (err) {
                error.NotMapped => not_mapped = true,
            };
        }

        if (not_mapped)
            return error.NotMapped;
    }

    const SetPermsError = PageTable.SetPermsError || error{NotUserRange};

    pub fn setRangePerms(
        self: *AddressSpace,
        addr: usize,
        length: usize,
        perms: Perms,
    ) SetPermsError!void {
        try checkRange(addr, length);

        // Set permissions for each page
        const addr_end = addr + length;
        var page_base: usize = addr;
        while (page_base < addr_end) : (page_base += std.mem.page_size) {
            try self.page_table.setPagePerms(page_base, perms);
        }
    }

    pub fn clone(self: AddressSpace) !AddressSpace {
        return AddressSpace{
            .allocator = self.allocator,
            .page_table = try PageTable.clone(self.page_table),
        };
    }
};
